src/librustc/dep_graph/graph.rs

```diff
@@ -948,8 +939,10 @@ struct DepNodeData {
 }
 
 pub(super) struct CurrentDepGraph {
-    data: IndexVec<DepNodeIndex, DepNodeData>,
-    node_to_node_index: FxHashMap<DepNode, DepNodeIndex>,
+    data: Lock<IndexVec<DepNodeIndex, DepNodeData>>,
```

Comment:
The comment below is great -- but it'd be good to document just a bit more (this is pre-existing, as the code wasn't all that thoroughlly commented before =)

I'd say something like:

---

Encodes the rustc dependency graph. The nodes are identified by an index (`DepNodeIndex`); the data for each node is stored in its `DepNodeData`, found in the `data` field.

We never remove nodes from the graph: they are only added. Most of the time, the node is mapped 1-to-1 with some `DepNode`, which basically identifies a query. When such nodes are allocated, we add the `DepNod` into the `node_to_node_index` map and allocate a fresh node index.

This struct uses two locks internally: the `data` and `node_to_node_index` field are locked separately. Operations that begin with a `DepNodeIndex` typically just access the data field. 

The only operation that must manipulate both fields is adding new nodes, in which case we first acquire the `node_to_node_index` lock and then, once a new node is to be inserted, acquire the lock on `data.`

(We also have a field fields that use atomics: these are simple counters that are used for profiling and debugging and are not used with `debug_assertions`.)

Reply:
I added that, but removed the incorrect parts =P